pkg/resizer/csi_resizer.go

```diff
@@ -0,0 +1,162 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resizer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/csi"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+// NewCSIResizer creates a new resizer responsible for resizing CSI volumes.
+func NewCSIResizer(address string, timeout time.Duration) (Resizer, error) {
+	client, err := csi.New(address, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("connect to CSI driver failed: %v", err)
+	}
+
+	if err := waitDriverReady(client, timeout); err != nil {
+		return nil, err
+	}
+
+	name, err := getDriverName(client, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("get driver name failed: %v", err)
+	}
+
+	if err := supportsPluginControllerService(client, timeout); err != nil {
+		return nil, err
+	}
+
+	if err := supportsControllerResize(client, timeout); err != nil {
+		return nil, err
+	}
+
+	return &csiResizer{
+		name:    name,
+		client:  client,
+		timeout: timeout,
+	}, nil
+}
+
+type csiResizer struct {
+	name    string
+	client  csi.Client
+	timeout time.Duration
+}
+
+func (r *csiResizer) Name() string {
+	return r.name
+}
+
+func (r *csiResizer) CanSupport(pv *v1.PersistentVolume) bool {
+	source := pv.Spec.CSI
+	if source == nil {
+		klog.V(4).Infof("PV %s is not a CSI volume, skip it", pv.Name)
+		return false
+	}
+	if source.Driver != r.name {
+		klog.V(4).Infof("Skip resize PV %s for resizer %s", pv.Name, source.Driver)
+		return false
+	}
+	return true
+}
+
+func (r *csiResizer) Resize(pv *v1.PersistentVolume, requestSize resource.Quantity) (resource.Quantity, bool, error) {
+	oldSize := pv.Spec.Capacity[v1.ResourceStorage]
+
+	source := pv.Spec.CSI
+	if source == nil {
+		return oldSize, false, errors.New("not a CSI volume")
+	}
+	volumeID := source.VolumeHandle
+	if len(volumeID) == 0 {
+		return oldSize, false, errors.New("empty volume handle")
+	}
+
+	ctx, cancel := timeoutCtx(r.timeout)
+	defer cancel()
+	// TODO: Get secrets used for expansion operation.
+	newSizeBytes, nodeResizeRequired, err := r.client.Expand(ctx, volumeID, requestSize.Value(), nil)
+	if err != nil {
+		return oldSize, nodeResizeRequired, err
+	}
+	return *resource.NewQuantity(newSizeBytes, resource.BinarySI), nodeResizeRequired, err
+}
+
+func waitDriverReady(client csi.Client, timeout time.Duration) error {
+	err := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		ctx, cancel := timeoutCtx(timeout)
+		defer cancel()
+		if err := client.Probe(ctx); err != nil {
+			klog.V(4).Infof("Driver not ready: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout {
+			return errors.New("waiting for driver ready timeout")
+		}
+		return err
+	}
+	return nil
+}
+
+func getDriverName(client csi.Client, timeout time.Duration) (string, error) {
```

Comment:
I would preferably convert these functions to directly accept contexts rather than converting them around.